src/realm/impl/transact_log.hpp

```diff
@@ -793,119 +838,150 @@ inline char* TransactLogEncoder::encode_double(char* ptr, double value)
     return std::copy_n(val_ptr, sizeof value, ptr);
 }
 
-template <class T>
-struct TransactLogEncoder::EncodeNumber {
-    void operator()(T value, char** ptr)
-    {
-        auto value_2 = value + 0; // Perform integral promotion
-        *ptr = encode_int(*ptr, value_2);
-    }
-};
 template <>
-struct TransactLogEncoder::EncodeNumber<char> {
-    void operator()(char value, char** ptr)
-    {
-        // Write the char as-is without encoding.
-        **ptr = value;
-        ++(*ptr);
-    }
-};
+inline char* TransactLogEncoder::encode<DataType>(char* ptr, DataType type)
+{
+    return encode<char>(ptr, type);
+}
+
 template <>
-struct TransactLogEncoder::EncodeNumber<bool> {
-    void operator()(bool value, char** ptr)
-    {
-        *ptr = encode_bool(*ptr, value);
-    }
-};
+inline char* TransactLogEncoder::encode<StringData>(char* ptr, StringData s)
+{
+    ptr = encode_int(ptr, s.size());
+    return std::copy_n(s.data(), s.size(), ptr);
+}
+
 template <>
-struct TransactLogEncoder::EncodeNumber<float> {
-    void operator()(float value, char** ptr)
-    {
-        *ptr = encode_float(*ptr, value);
+inline char* TransactLogEncoder::encode<TransactLogEncoder::IntegerList>(char* ptr,
+                                                                         TransactLogEncoder::IntegerList list)
+{
+    auto i = std::get<0>(list);
+    auto end = std::get<1>(list);
```

Comment:
Would improve readability *alot* if using `IntegerColumnIterator` instead of `auto`. In the code that follows, and in the method arguments, nothing hints/indicates its type, so you'd really need to reverse engineer a bit to figure it out